### PR TITLE
[PSR-7] Bridge documentation

### PR DESCRIPTION
**Pull request by dunglas**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes |
| Applies to | 2.7 symfony/psr-http-message-bridge#1 |
| Fixed tickets | n/a |

Documents the PSR-7 Bridge. As there was no previous doc for bridges, I created a structure but not sur if this is a good idea.

### CODE REVIEW

**Comment by WouterJ** [comment]
I think it's better to just have a new cookbook section about PSR-7 and how to use it in your application.

File: bridges/psr7.rst [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+Diactoros :class:`Zend\\Diactoros\\Response` implementing the :class:`Psr\Http\Message\ResponseInterface`
+interface::
+
+    use Symfony\Bridge\PsrHttpMessage\Factory\DiactorosFactory;
+    use Symfony\Component\HttpFoundation\Response;
+
+    $symfonyResponse = new Response('Content');
+
+    $psr7Factory = new DiactorosFactory();
+    $psrRequest = $psr7Factory->createResponse($symfonyResponse);
```
Comment (weaverryan) [inline bridges/psr7.rst]: $psrResponse =

**Comment by dunglas** [comment]
Moved to a cookbook entry.

File: cookbook/psr7.rst [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    use Symfony\Component\HttpFoundation\Request;
+
+    $symfonyRequest = new Request(array(), array(), array(), array(), array(), array('HTTP_HOST' => 'dunglas.fr'), 'Content');
+    // The HTTP_HOST server key must be set to avoid an unexpected error
+
+    $psr7Factory = new DiactorosFactory();
+    $psrRequest = $psr7Factory->createRequest($symfonyRequest);
+
+And now from a :class:`Symfony\\Component\\HttpFoundation\\Response` to a Zend
+Diactoros :class:`Zend\\Diactoros\\Response` implementing the :class:`Psr\Http\Message\ResponseInterface`
```
Comment (xabbuh) [inline cookbook/psr7.rst]: Backslashes need to be escaped for the `ResponseInterface` too.

**Comment by dunglas** [comment]
@xabbuh done

File: cookbook/map.rst.inc [hunk 3]
```diff
@@ -195,6 +195,10 @@
   * (configuration) :doc:`/cookbook/configuration/pdo_session_storage`
   * :doc:`/cookbook/session/avoid_session_start`
 
+* **PSR-7** *
```
Comment (xabbuh) [inline cookbook/map.rst.inc]: The last `*` needs to be removed.

File: cookbook/psr7.rst [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+The bridge also needs a PSR-7 implementation to allow converting HttpFoundation
+objects to PSR-7 objects. It provides native support for _`Zend Diactoros`_.
+Use Composer (``zendframework/zend-diactoros`` on `Packagist`_) or refers to
+the project documentation to install it.
+
+Usage
+-----
+
+Converting from HttpFoundation objects to PSR-7
```
Comment (xabbuh) [inline cookbook/psr7.rst]: Objects

File: cookbook/psr7.rst [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+    use Symfony\Bridge\PsrHttpMessage\Factory\DiactorosFactory;
+    use Symfony\Component\HttpFoundation\Response;
+
+    $symfonyResponse = new Response('Content');
+
+    $psr7Factory = new DiactorosFactory();
+    $psrResponse = $psr7Factory->createResponse($symfonyResponse);
+
+Converting objects implementing PSR-7 interfaces to HttpFoundation
```
Comment (xabbuh) [inline cookbook/psr7.rst]: [...] Objects [...] Interfaces [...]

File: cookbook/psr7.rst [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+On the other hand, the bridge provide a factory interface called :class:`Symfony\\Bridge\\PsrHttpMessage\\HttpFoundationFactoryInterface`
```
Comment (xabbuh) [inline cookbook/psr7.rst]: Please move the `class` role to its own line.

File: cookbook/psr7.rst [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+that builds HttpFoundation objects from objects implementing PSR-7 interfaces.
+
+The next snippet explain how to convert an object implementing the :class:`Psr\\Http\\Message\\ServerRequestInterface`
```
Comment (xabbuh) [inline cookbook/psr7.rst]: same here

**Comment by xabbuh** [comment]
We also need an entry for the new document in the toctree in `/cookbook/index.rst`.

**Comment by dunglas** [comment]
Should be ok now.

**Comment by harikt** [comment]
I have a question : Does Symfony 3.0 http foundation will make use of psr-7 or it will be using the bridge itself ?

I have noticed some issues regarding splitting of the session component from http foundation etc.

Thank you.

**Comment by weaverryan** [comment]
@harikt 3.0 will not use psr-7 (so you'd use the bridge like now). We need to protect backwards compatibility - but we are discussing the best (and most responsible way) to move towards psr-7.

**Comment by weaverryan** [comment]
Awesome @dunglas  - thanks!

**Pull request merged**